querydsl-core/src/main/java/com/querydsl/core/types/dsl/NumberExpression.java

```diff
@@ -203,9 +203,11 @@ private T cast(Number number) {
         return ceil;
     }
 
-    private Class<?> getDivisionType(Class<?> left, Class<?> right) {
+    private <N> Class<? extends N> getDivisionType(Class<N> left, Class<?> right) {
         if (!left.equals(right)) {
-            return Double.class;
+            @SuppressWarnings("unchecked") // This is a good in the middle decision
+            Class<? extends N> newType = (Class<? extends N>) Double.class;
+            return newType;
```

Comment:
I am not sure about this method, it's now safer, doesn't introduce a new signature (erasure is the same) but what do you think?
@timowest @johnktims